CGI discovery phase changes

cgiHarland opened the pull request:
Contains a model update.

Discovery phase changes - DX number moved, Court code search added, Areas of Law external URL added and breadcrumbs updated.

DX number - The 'Legal professionals' block has been removed and the DX number and court number moved to the 'Visit Us' block.
Court code - A new search option has been added to allow users to search specifically on the court number; breadcrumbs have also been updated.
AoL - Links have been applied to these with external links to Government sites. This involves a model update!
Breadcrumbs - The 'Court list' crumb has been changed to 'A-Z list of all courts'

richerry commented on courtfinder/courts/templates/courts/court.jinja:
```diff
@@ -8,21 +8,26 @@
     <li><a href="{% url 'staticpages:home' %}">Home</a></li>
     <li><a href="{% url 'search:search' %}">Find a court or tribunal</a></li>
-    {% if query %}
-    <li><a href="{% url 'search:results' %}">Search by name of address</a></li>
```
Indentation.

lgarvey commented:
@cgiHarland I have run the tests and whilst they passed in master, they fail on your branch with a migration issue.  Also I see you have added no new tests to cover the changes made. Is that intentional?

richerry commented on courtfinder/search/court_search.py:
```diff
@@ -190,6 +198,23 @@ def __address_search( self, query ):
+        court_number_results =  sorted(Court.objects.filter(number__iregex=query_regex))
+
+        # put it all together and remove duplicates
+        results = list(OrderedDict.fromkeys(chain(court_number_results)))
```
Would this not have the same effect?

results = list(court_number_results)

Since you only iterate over it with list comprehension next you could maybe get away with just:

return [ result for result in court_number_results if result.displayed]

richerry commented on courtfinder/search/models.py:
```diff
@@ -58,9 +59,14 @@ def __unicode__(self):
 
 class AreaOfLaw(models.Model):
     name = models.CharField(max_length=255)
+    external_link = models.CharField(null=True, max_length=2048)
```
Consider blank=True as well if this is an optional field.
Max length of a url is 2083 characters.
Consider the Django URLField type for URLs.

richerry commented on courtfinder/search/models.py:
```diff
@@ -58,9 +59,14 @@ def __unicode__(self):
 class AreaOfLaw(models.Model):
     name = models.CharField(max_length=255)
+    external_link = models.CharField(null=True, max_length=2048)
+    external_link_desc = models.CharField(null=True, max_length=255)
```
blank=True maybe?

richerry commented on courtfinder/search/models.py:
```diff
@@ -58,9 +59,14 @@ def __unicode__(self):
+    external_link_desc = models.CharField(null=True, max_length=255)
 
+    def display_url(self): 
+        return urllib.unquote(self.external_link)
```
Check out what you get for free with the DjangoURL field, it also normalises to unicode by default!

richerry commented on courtfinder/search/models.py:
```diff
@@ -58,9 +59,14 @@ def __unicode__(self):
-        return self.name
+        return "%s: %s %s" % (self.name,  self.external_link, self.external_link_desc)
 
     class Meta:
```
Ordering a single field?

richerry commented on courtfinder/search/templates/search/courtcode.jinja:
```diff
@@ -0,0 +1,44 @@
+{% extends "layout/base.jinja" %}
+
+{% block page_title %}Search by Court code{% endblock %}
```
Internationalised user-facing strings?

cgiHarland replied:
See Lyndon's previous comment.

richerry commented on courtfinder/search/views.py:
```diff
@@ -11,6 +11,7 @@
 from search.models import Court, AreaOfLaw, DataStatus
 from search.court_search import CourtSearch, CourtSearchError, CourtSearchClientError, CourtSearchInvalidPostcode
 from search.rules import Rules
+from urlparse import urlparse
```
unused import

richerry commented on courtfinder/search/views.py:
```diff
@@ -11,6 +11,7 @@
+from urlparse import urlparse
 
 areas_of_law_description = {
     "Adoption": "applying to adopt a child.",
```
Internationalised user-facing strings?

cgiHarland replied:
See Lyndon's previous comment.

richerry commented on courtfinder/search/views.py:
```diff
@@ -72,24 +73,45 @@ def address(request):
+    courtcode = request.GET.get('courtcode', None)
+    
+    if courtcode is not None:
+        courtcode = re.sub(r'\s+',' ',courtcode.strip())
```
Consider compiling the regex at the module level. I've seen other compiled regexes in this review so not sure why we're not doing it for this one.

richerry commented on courtfinder/search/views.py:
```diff
@@ -72,24 +73,45 @@ def address(request):
+    courtcode = request.GET.get('courtcode', None)
+    return render(request, 'search/courtcode.jinja', {'error': error, 'courtcode' : courtcode})
+
 def results(request):
```
This function is poorly named, poorly documented and in danger of becoming too large.

cgiHarland replied:
Agree but outside of scope.

richerry commented on courtfinder/search/views.py:
```diff
@@ -72,24 +73,45 @@ def address(request):
+def courtcode(request):
+    error = request.GET.get('error', None)
+    courtcode = request.GET.get('courtcode', None)
+    return render(request, 'search/courtcode.jinja', {'error': error, 'courtcode' : courtcode})
```
Should we look up an internationalised user-facing string here instead of using whatever was passed in by the user?

cgiHarland replied:
See Lyndon's previous comment.

richerry commented on courtfinder/search/views.py:
```diff
@@ -72,24 +73,45 @@ def address(request):
+        if courtcode == '':
+            return redirect(reverse('search:courtcode')+'?error=noquery')
+        else:
+            results = CourtSearch(query=courtcode, courtcode_search=True).get_courts()
```
DANGER! 

Potential SQL injection vulnerability.

There is no validation of courtcode before using it as a query. I haven't followed the codepath beyond this call but it is bad practice and will help spread bad practices. Why are we not generally using Django form validation to validate django forms?

cgiHarland replied:
Doesn't the Django ORM take care of this?

richerry replied:
Django Form Validation is the tool for validating user input. The ORM will prevent a booby tables scenario from a string field but not a user putting what they want in the field without validation  and potentially seeing data they shouldn't.

lgarvey replied:
This pattern is repeated all over the courtfinder-search codebase.   Escaping happens at the db driver level and I'd trust that over a self-concocted sanitise function.

However, there's certainly merit to using forms to constrain values - but how would say a forms.CharField(max_length=100) provide any assistance with SQL injection? 

I don't see anywhere where unescaped values are being used with Django ORM RAW sql queries, so I'd regard this as a none issue.

richerry commented on courtfinder/search/views.py:
```diff
@@ -72,24 +73,45 @@ def address(request):
+            results = CourtSearch(query=courtcode, courtcode_search=True).get_courts()
 
-    if query is not None:
+            if len(results) > 0:
```
Indentation!

richerry commented on courtfinder/search/views.py:
```diff
@@ -72,24 +73,45 @@ def address(request):
+                    'search_results': __format_results(results)
+                })
+            else:
+                return redirect(reverse('search:courtcode')+'?error=noresults&q='+courtcode)
```
PEP8

richerry commented on courtfinder/search/views.py:
```diff
@@ -72,24 +73,45 @@ def address(request):
+            else:
+                return redirect(reverse('search:courtcode')+'?error=noresults&q='+courtcode)
+    elif query is not None:
         query = re.sub(r'\s+',' ',query.strip())
```
Compile at module level?

cgiHarland commented:
Hi Richard,
Which documentation are you talking about?

I’m knew to Github how do I respond to the comments?

Regards,
Andrew

Andrew Harland
Application Management Services | CGI
CGI, Hughes Park, Waterton Cross Business Park, South Road, Bridgend, CF31 3UL | United Kingdom
T: +44 (0) 845 070 7765 | M: +44 (0) 7500052221
andrew.harland@cgi.com<mailto:andrew.harland@cgi.com> | cgi-group.co.uk<http://www.cgi-group.co.uk/>

From: richerry [mailto:notifications@github.com]
Sent: 06 January 2017 12:38
To: ministryofjustice/courtfinder-search
Cc: Harland, Andrew; Mention
Subject: Re: [ministryofjustice/courtfinder-search] Based on master! (#151)


@richerry requested changes on this pull request.

Further to the individual comments I see no changes to tests or documentation.

________________________________

In courtfinder/courts/templates/courts/court.jinja<https://github.com/ministryofjustice/courtfinder-search/pull/151#pullrequestreview-15476496>:

>      {% elif postcode %}

-    <li><a href="{% url 'search:aol' %}?aol={{aol}}">About your issue</a></li>

-    {%   if spoe %}

spoe is a poor choice of name.

________________________________

In courtfinder/courts/templates/courts/court.jinja<https://github.com/ministryofjustice/courtfinder-search/pull/151#pullrequestreview-15476496>:

> @@ -8,21 +8,26 @@

   <ol role="breadcrumbs" class="group">

     <li><a href="{% url 'staticpages:home' %}">Home</a></li>

     <li><a href="{% url 'search:search' %}">Find a court or tribunal</a></li>

-    {% if query %}

-    <li><a href="{% url 'search:results' %}">Search by name of address</a></li>

Indentation

________________________________

In courtfinder/courts/templates/courts/court.jinja<https://github.com/ministryofjustice/courtfinder-search/pull/151#pullrequestreview-15476496>:

> @@ -66,7 +71,20 @@

         {{ court.postal_address.county }}<br/>

         {% endif %}

         {{ court.postal_address.postcode }}

-

+      </div>

+      {% endif %}

+      {% if court.number or court.dx_number %}

+      <div id="pros">

+        <dl>

+          {% if court.number %}

+          <dt>Court/tribunal no.</dt>

Indentation

________________________________

In courtfinder/courts/templates/courts/court.jinja<https://github.com/ministryofjustice/courtfinder-search/pull/151#pullrequestreview-15476496>:

> @@ -66,7 +71,20 @@

         {{ court.postal_address.county }}<br/>

         {% endif %}

         {{ court.postal_address.postcode }}

-

+      </div>

+      {% endif %}

+      {% if court.number or court.dx_number %}

+      <div id="pros">

+        <dl>

+          {% if court.number %}

+          <dt>Court/tribunal no.</dt>

+          <dd>{{ court.number }}</dd>

+          {% endif %}

+          {% if court.dx_number %}

+          <dt><abbr title="document exchange">DX</abbr>:</dt>

+          <dd>{{ court.dx_number }}</dd>

+          {% endif %}

+        </dl>

       </div>

       {% endif %}

Should there be anything displayed to the user if there is no court number or dx number?

________________________________

In courtfinder/courts/templates/courts/court.jinja<https://github.com/ministryofjustice/courtfinder-search/pull/151#pullrequestreview-15476496>:

> @@ -8,21 +8,26 @@

   <ol role="breadcrumbs" class="group">

     <li><a href="{% url 'staticpages:home' %}">Home</a></li>

     <li><a href="{% url 'search:search' %}">Find a court or tribunal</a></li>

-    {% if query %}

-    <li><a href="{% url 'search:results' %}">Search by name of address</a></li>

-    <li><a href="{% url 'search:results' %}?q={{ query }}">Search results</a></li>

+    {% if query or courtcode %}

+      {% if query %}

+        <li><a href="{% url 'search:address' %}">Search by name of address</a></li>

Should this be 'name or address'? I've never heard of a 'name of address'

________________________________

In courtfinder/courts/templates/courts/court.jinja<https://github.com/ministryofjustice/courtfinder-search/pull/151#pullrequestreview-15476496>:

> @@ -8,21 +8,26 @@

   <ol role="breadcrumbs" class="group">

     <li><a href="{% url 'staticpages:home' %}">Home</a></li>

How have we handled internationalised user-facing strings?

________________________________

In courtfinder/search/court_search.py<https://github.com/ministryofjustice/courtfinder-search/pull/151#pullrequestreview-15476496>:

> @@ -38,7 +38,8 @@ class CourtSearchInvalidPostcode(CourtSearchError):



 class CourtSearch:



-    def __init__( self, postcode=None, area_of_law=None, single_point_of_entry=False, query=None ):

+    def __init__( self, postcode=None, area_of_law=None, single_point_of_entry=False, query=None, courtcode_search=False ):

Python code should pass PEP8 and lint (pylint/pyflakes) checks. This line is too long.

________________________________

In courtfinder/search/court_search.py<https://github.com/ministryofjustice/courtfinder-search/pull/151#pullrequestreview-15476496>:

> @@ -38,7 +38,8 @@ class CourtSearchInvalidPostcode(CourtSearchError):



 class CourtSearch:

Can we have a docstring please?

And I'm not sure that a class makes sense here. Lint checkers would also warn about too few methods. Is this not a candidate for a function instead?

________________________________

In courtfinder/search/court_search.py<https://github.com/ministryofjustice/courtfinder-search/pull/151#pullrequestreview-15476496>:

> @@ -38,7 +38,8 @@ class CourtSearchInvalidPostcode(CourtSearchError):



 class CourtSearch:



-    def __init__( self, postcode=None, area_of_law=None, single_point_of_entry=False, query=None ):

+    def __init__( self, postcode=None, area_of_law=None, single_point_of_entry=False, query=None, courtcode_search=False ):

+        self.courtcode_search = courtcode_search

         if query:

Should be:

if query is not None:

________________________________

In courtfinder/search/court_search.py<https://github.com/ministryofjustice/courtfinder-search/pull/151#pullrequestreview-15476496>:

> @@ -60,7 +61,14 @@ def __init__( self, postcode=None, area_of_law=None, single_point_of_entry=False



     def get_courts( self ):

         if hasattr(self, 'query'):

-            return self.__address_search(self.query)

+            numberRE = re.compile('^[0-9]{3,4}$')

Regexes should be compiled once when the app loads, not every time it is used.

________________________________

In courtfinder/search/court_search.py<https://github.com/ministryofjustice/courtfinder-search/pull/151#pullrequestreview-15476496>:

> @@ -60,7 +61,14 @@ def __init__( self, postcode=None, area_of_law=None, single_point_of_entry=False



     def get_courts( self ):

         if hasattr(self, 'query'):

-            return self.__address_search(self.query)

+            numberRE = re.compile('^[0-9]{3,4}$')

+            courtNumber = numberRE.match(self.query)

+

+            if self.courtcode_search == True: #courtNumber:

PEP 8.

________________________________

In courtfinder/search/court_search.py<https://github.com/ministryofjustice/courtfinder-search/pull/151#pullrequestreview-15476496>:

> @@ -190,6 +198,23 @@ def __address_search( self, query ):



         return [result for result in results if result.displayed]



+    def __court_number_search( self, query ):

+        """

+        Retrieve court(s) by court code, order and remove duplicates

+        """

+

+        # First we get courts whose court code contains the query string

+        # (for these courts sorted to show the courts with the highest number of areas of law first)

+

+        word_separator = re.compile(r'[^\w]+', re.UNICODE)

Compile static regex patterns once, not every time they are used.

________________________________

In courtfinder/search/court_search.py<https://github.com/ministryofjustice/courtfinder-search/pull/151#pullrequestreview-15476496>:

> @@ -190,6 +198,23 @@ def __address_search( self, query ):



         return [result for result in results if result.displayed]



+    def __court_number_search( self, query ):

+        """

+        Retrieve court(s) by court code, order and remove duplicates

+        """

+

+        # First we get courts whose court code contains the query string

+        # (for these courts sorted to show the courts with the highest number of areas of law first)

+

+        word_separator = re.compile(r'[^\w]+', re.UNICODE)

+        query_regex = ''.join(map(lambda word: "(?=.*\y"+word+"\y)", re.split(word_separator, query)))

+

+        court_number_results =  sorted(Court.objects.filter(number__iregex=query_regex))

Might be better to sort using Django's order_by() method as part of the query.

________________________________

In courtfinder/search/court_search.py<https://github.com/ministryofjustice/courtfinder-search/pull/151#pullrequestreview-15476496>:

> @@ -190,6 +198,23 @@ def __address_search( self, query ):



         return [result for result in results if result.displayed]



+    def __court_number_search( self, query ):

+        """

+        Retrieve court(s) by court code, order and remove duplicates

+        """

+

+        # First we get courts whose court code contains the query string

+        # (for these courts sorted to show the courts with the highest number of areas of law first)

+

+        word_separator = re.compile(r'[^\w]+', re.UNICODE)

+        query_regex = ''.join(map(lambda word: "(?=.*\y"+word+"\y)", re.split(word_separator, query)))

+

+        court_number_results =  sorted(Court.objects.filter(number__iregex=query_regex))

+

+        # put it all together and remove duplicates

+        results = list(OrderedDict.fromkeys(chain(court_number_results)))

Would this not have the same effect?

results = list(court_number_results)

Since you only iterate over it with list comprehension next you could maybe get away with just:

return [ result for result in court_number_results if result.displayed]

________________________________

In courtfinder/search/models.py<https://github.com/ministryofjustice/courtfinder-search/pull/151#pullrequestreview-15476496>:

> @@ -58,9 +59,14 @@ def __unicode__(self):



 class AreaOfLaw(models.Model):

     name = models.CharField(max_length=255)

+    external_link = models.CharField(null=True, max_length=2048)

Consider blank=True as well if this is an optional field.
Max length of a url is 2083 characters.
Consider the Django URLField type for URLs.

________________________________

In courtfinder/search/models.py<https://github.com/ministryofjustice/courtfinder-search/pull/151#pullrequestreview-15476496>:

> @@ -58,9 +59,14 @@ def __unicode__(self):



 class AreaOfLaw(models.Model):

     name = models.CharField(max_length=255)

+    external_link = models.CharField(null=True, max_length=2048)

+    external_link_desc = models.CharField(null=True, max_length=255)

blank=True maybe?

________________________________

In courtfinder/search/models.py<https://github.com/ministryofjustice/courtfinder-search/pull/151#pullrequestreview-15476496>:

>

+    def display_url(self):

+        return urllib.unquote(self.external_link)

Check out what you get for free with the DjangoURL field, it also normalises to unicode by default!

________________________________

In courtfinder/search/models.py<https://github.com/ministryofjustice/courtfinder-search/pull/151#pullrequestreview-15476496>:

>      def __unicode__(self):

-        return self.name

+        return "%s: %s %s" % (self.name,  self.external_link, self.external_link_desc)



     class Meta:

Ordering a single field?

________________________________

In courtfinder/search/templates/search/courtcode.jinja<https://github.com/ministryofjustice/courtfinder-search/pull/151#pullrequestreview-15476496>:

> @@ -0,0 +1,44 @@

+{% extends "layout/base.jinja" %}

+

+{% block page_title %}Search by Court code{% endblock %}

Internationalised user-facing strings?

________________________________

In courtfinder/search/views.py<https://github.com/ministryofjustice/courtfinder-search/pull/151#pullrequestreview-15476496>:

> @@ -11,6 +11,7 @@

 from search.models import Court, AreaOfLaw, DataStatus

 from search.court_search import CourtSearch, CourtSearchError, CourtSearchClientError, CourtSearchInvalidPostcode

 from search.rules import Rules

+from urlparse import urlparse

unused import

________________________________

In courtfinder/search/views.py<https://github.com/ministryofjustice/courtfinder-search/pull/151#pullrequestreview-15476496>:

> @@ -11,6 +11,7 @@

 from search.models import Court, AreaOfLaw, DataStatus

 from search.court_search import CourtSearch, CourtSearchError, CourtSearchClientError, CourtSearchInvalidPostcode

 from search.rules import Rules

+from urlparse import urlparse



 areas_of_law_description = {

     "Adoption": "applying to adopt a child.",

Internationalised user-facing strings?

________________________________

In courtfinder/search/views.py<https://github.com/ministryofjustice/courtfinder-search/pull/151#pullrequestreview-15476496>:

>  def results(request):

     query = request.GET.get('q', None)

+    courtcode = request.GET.get('courtcode', None)

+

+    if courtcode is not None:

+        courtcode = re.sub(r'\s+',' ',courtcode.strip())

Consider compiling the regex at the module level. I've seen other compiled regexes in this review so not sure why we're not doing it for this one.

________________________________

In courtfinder/search/views.py<https://github.com/ministryofjustice/courtfinder-search/pull/151#pullrequestreview-15476496>:

> @@ -72,24 +73,45 @@ def address(request):

     query = request.GET.get('q', None)

     return render(request, 'search/address.jinja', {'error': error, 'query':query})



+def courtcode(request):

+    error = request.GET.get('error', None)

+    courtcode = request.GET.get('courtcode', None)

+    return render(request, 'search/courtcode.jinja', {'error': error, 'courtcode' : courtcode})

+

 def results(request):

This function is poorly named, poorly documented and in danger of becoming too large.

________________________________

In courtfinder/search/views.py<https://github.com/ministryofjustice/courtfinder-search/pull/151#pullrequestreview-15476496>:

> @@ -72,24 +73,45 @@ def address(request):

     query = request.GET.get('q', None)

     return render(request, 'search/address.jinja', {'error': error, 'query':query})



+def courtcode(request):

+    error = request.GET.get('error', None)

+    courtcode = request.GET.get('courtcode', None)

+    return render(request, 'search/courtcode.jinja', {'error': error, 'courtcode' : courtcode})

Should we look up an internationalised user-facing string here instead of using whatever was passed in by the user?

________________________________

In courtfinder/search/views.py<https://github.com/ministryofjustice/courtfinder-search/pull/151#pullrequestreview-15476496>:

>  def results(request):

     query = request.GET.get('q', None)

+    courtcode = request.GET.get('courtcode', None)

+

+    if courtcode is not None:

+        courtcode = re.sub(r'\s+',' ',courtcode.strip())

+        if courtcode == '':

+            return redirect(reverse('search:courtcode')+'?error=noquery')

+        else:

+            results = CourtSearch(query=courtcode, courtcode_search=True).get_courts()

DANGER!

Potential SQL injection vulnerability.

There is no validation of courtcode before using it as a query. I haven't followed the codepath beyond this call but it is bad practice and will help spread bad practices. Why are we not generally using Django form validation to validate django forms?

________________________________

In courtfinder/search/views.py<https://github.com/ministryofjustice/courtfinder-search/pull/151#pullrequestreview-15476496>:

>

-    if query is not None:

+            if len(results) > 0:

Indentation!

________________________________

In courtfinder/search/views.py<https://github.com/ministryofjustice/courtfinder-search/pull/151#pullrequestreview-15476496>:

>

-    if query is not None:

+            if len(results) > 0:

+                return render(request, 'search/results.jinja', {

+                    'query': courtcode,

+                    'courtcode_search': True,

+                    'search_results': __format_results(results)

+                })

+            else:

+                return redirect(reverse('search:courtcode')+'?error=noresults&q='+courtcode)

PEP8

________________________________

In courtfinder/search/views.py<https://github.com/ministryofjustice/courtfinder-search/pull/151#pullrequestreview-15476496>:

>

-    if query is not None:

+            if len(results) > 0:

+                return render(request, 'search/results.jinja', {

+                    'query': courtcode,

+                    'courtcode_search': True,

+                    'search_results': __format_results(results)

+                })

+            else:

+                return redirect(reverse('search:courtcode')+'?error=noresults&q='+courtcode)

+    elif query is not None:

         query = re.sub(r'\s+',' ',query.strip())

Compile at module level?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/ministryofjustice/courtfinder-search/pull/151#pullrequestreview-15476496>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AWEJCpHUp9xZ2TnNfiYfC90brL_3nVpeks5rPjWzgaJpZM4Lchw0>.

lgarvey commented on courtfinder/courts/templates/courts/court.jinja:
```diff
@@ -8,21 +8,26 @@
   <ol role="breadcrumbs" class="group">
     <li><a href="{% url 'staticpages:home' %}">Home</a></li>
```
There's currently no i18n on courtfinder and adding it is out of scope for the changes being made by CGI.  So happy to allow this.

lgarvey commented on courtfinder/courts/templates/courts/court.jinja:
```diff
@@ -8,21 +8,26 @@
+      {% endif %}
     {% elif postcode %}
-    <li><a href="{% url 'search:aol' %}?aol={{aol}}">About your issue</a></li>
-    {%   if spoe %}
```
This variable was already named spoe - also the URL generation is ugly and poorly designed but it was already working this way. So I'd regard this as out of scope again for the changes CGI have been enlisted to make.

lgarvey commented on courtfinder/search/court_search.py:
```diff
@@ -38,7 +38,8 @@ class CourtSearchInvalidPostcode(CourtSearchError):
 class CourtSearch:
 
-    def __init__( self, postcode=None, area_of_law=None, single_point_of_entry=False, query=None ):
+    def __init__( self, postcode=None, area_of_law=None, single_point_of_entry=False, query=None, courtcode_search=False ):
```
If you're talking PEP8 compliance then the white space either side of the arg list will also need removing.

There's also a telescopic constructor anti pattern which would could be fixed by refactoring the class so that it provides a fluent interface, but I don't expect CGI to do this work.

I'd suggest @richerry it could be a worthwhile exercise for you to fix these aspects of the codebase.

lgarvey commented on courtfinder/search/court_search.py:
```diff
@@ -38,7 +38,8 @@ class CourtSearchInvalidPostcode(CourtSearchError):
-    def __init__( self, postcode=None, area_of_law=None, single_point_of_entry=False, query=None ):
+    def __init__( self, postcode=None, area_of_law=None, single_point_of_entry=False, query=None, courtcode_search=False ):
+        self.courtcode_search = courtcode_search
         if query:
```
Actually it's more pythonic to say: 

if query:

Unless you're specfically testing for None

richerry replied:
> Unless you're specfically testing for not None

Which in this case we are because None is the default value

lgarvey replied:
No, you're not interested in differentiating between None and another "falsey" value, such as False, in this instance.  So my comment still stands and the pythonic way is to write:

if not query

But it's such a minor detail that I would accept either form.

lgarvey commented on courtfinder/search/court_search.py:
```diff
@@ -60,7 +61,14 @@ def __init__( self, postcode=None, area_of_law=None, single_point_of_entry=False
+            numberRE = re.compile('^[0-9]{3,4}$')
+            courtNumber = numberRE.match(self.query)
+
+            if self.courtcode_search == True: #courtNumber:    
```
Pythonic solution is 

if self.courtcode_search:

cgiHarland replied:
Yep, modified in local branch awaiting to update pull request.

richerry replied:
If the courtcode_search were "" it would evaluate to False. I've no idea what possible values could be passed in. If we used :param foo: and :type foo: comments in the docstrings it would help consumers of the function, help with testing as well as provide a better understanding of the code.

Stuff like that would be out of scope though.

cgiHarland commented on courtfinder/courts/templates/courts/court.jinja:
```diff
@@ -8,21 +8,26 @@
-    <li><a href="{% url 'search:results' %}?q={{ query }}">Search results</a></li>
+    {% if query or courtcode %} 
+      {% if query %} 
+        <li><a href="{% url 'search:address' %}">Search by name of address</a></li>
```
Yep, modified in local branch awaiting to update pull request.

cgiHarland commented on courtfinder/courts/templates/courts/court.jinja:
```diff
@@ -66,7 +71,20 @@
+          {% endif %}
+        </dl>
       </div>
       {% endif %}
```
No - Previous code would have output 'Legal Professionals' if either a court number or DX was found. But we lost the 'Legal ...' title, as per the request. maybe raise with users.

cgiHarland commented on courtfinder/search/court_search.py:
```diff
@@ -38,7 +38,8 @@ class CourtSearchInvalidPostcode(CourtSearchError):
 
 class CourtSearch:
```
Maybe, but I think that's outside the scope of these changes.

cgiHarland commented on courtfinder/search/court_search.py:
```diff
@@ -60,7 +61,14 @@ def __init__( self, postcode=None, area_of_law=None, single_point_of_entry=False
     def get_courts( self ):
         if hasattr(self, 'query'):
-            return self.__address_search(self.query)
+            numberRE = re.compile('^[0-9]{3,4}$')
```
A good point, so I started to read about where they best place to put this would be and found out that that 'compile' uses a cache. As this is only used here I don't think we need to change this.

richerry replied:
I'd like to read more about this because I've never heard of it. I'm pretty sure that once the request/response goes out of scope the cached versions will be lost. A link I'll give to support the idea that they should be compiled at module scope is: http://stackoverflow.com/questions/65266/caching-compiled-regex-objects-in-python

lgarvey replied:
The regex appears to get used once per request. If so, then this is clearly an example of premature optimisation.  http://wiki.c2.com/?PrematureOptimization

Also I've been scanning the source code for the python re module and it has a module level LRU cache for frequently used regexes, whether they have been compiled or not.

So you could get rid of compile altogether and write this:

courtNumber = re.match('^[0-9]{3,4}$', self.query)

If it makes sense to do so.

cgiHarland commented on courtfinder/search/court_search.py:
```diff
@@ -190,6 +198,23 @@ def __address_search( self, query ):
+        # First we get courts whose court code contains the query string
+        # (for these courts sorted to show the courts with the highest number of areas of law first)
+
+        word_separator = re.compile(r'[^\w]+', re.UNICODE)
```
Compile uses a cache, so no need to pre-compile.

lgarvey commented on courtfinder/search/tests/test_courtcode.py:
```diff
@@ -0,0 +1,40 @@
+        Ingest.courts(json.loads(courts_json_1))
+        DataStatus.objects.create(data_hash='415d49233b8592cf5195b33f0eddbdc86cebc72f2d575d392e941a53c085281a')
+
+    def tearDown(self):
```
This method is unused. Can you remove it.

lgarvey commented on courtfinder/search/tests/test_courtcode.py:
```diff
@@ -0,0 +1,40 @@
+
+    def test_non_existing(self):
+        """
+        Only numerics
```
Can you add a proper description here? Or remove it

lgarvey commented on courtfinder/search/tests/test_courtcode.py:
```diff
@@ -0,0 +1,40 @@
+
+    def test_exists(self):
+        """
+        Only numerics
```
Same here
